Add push_matrix fcn to COO

CattleProdigy opened the pull request:
This MR adds a function which allow entire dense matrices to be pushed into a COO matrix. 

I wrote unit tests, but I haven't actually used it enough to recommend merging yet. I'm just hoping for any initial feedback.

CattleProdigy commented on nalgebra-sparse/tests/unit_tests/coo.rs:
```diff
@@ -252,3 +252,97 @@ fn coo_push_out_of_bounds_entries() {
+        );
+    }
+
+    let csr = nalgebra_sparse::csr::CsrMatrix::from(&coo);
```
axe this

CattleProdigy replied:
https://github.com/dimforge/nalgebra/pull/904/commits/6b1b9077fe76a2644501f07179a2d2e7823821ee

Andlon commented on nalgebra-sparse/src/coo.rs:
```diff
@@ -45,6 +45,43 @@ pub struct CooMatrix<T> {
+    /// Panics if any part of the dense matrix is out of bounds of the sparse matrix
+    /// when inserted at `(r, c)`.
+    #[inline]
+    pub fn push_mat<R: na::Dim, C: na::Dim, S: nalgebra::storage::Storage<T, R, C>>(
```
To add something more than just my personal taste here, the reason is that abbreviations tend to hurt readability a little because the more abbreviations you need to keep in the back of your head, the slower you are able to read the code. Of course, sometimes abbreviations can help code readability by reducing character counts, but at least in my opinion `push_matrix` does not meet this criterion!

CattleProdigy replied:
Contrary to what I used here, I strongly prefer unabbreviated identifiers when possible, so I'd happily go with `push_matrix`. Or even `push_dense_matrix`?

CattleProdigy replied:
https://github.com/dimforge/nalgebra/pull/904/commits/fec32b1de2d95faa7509359326ce2ab89981d489

Andlon replied:
I agree that `push_dense_matrix` is even more precise. I think perhaps we can get away with `push_matrix` here though, since the canonical matrix type in `nalgebra` is still `Matrix`, and so sparse formats rather represent the exceptional case. Also, when reading the code, it doesn't *really* matter if the matrix would be dense or sparse, as mathematically they would accomplish the same thing - so the inferred logic behind the code would be the same. I think my vote goes to `push_matrix`.

Andlon reviewed (changes requested):
Hey @CattleProdigy, thanks a lot for this PR! This is a nice contribution to the library, and I think the PR is very clean. My only tidbit is that I think I'd like `push_matrix` instead of the abbreviated `push_mat` (see comment). What do you think?

CattleProdigy commented:
(force pushed to fix the commit author email)

Andlon commented:
@CattleProdigy: I'm a little confused of the state right now because of the rebase and the fixup commit messages. Is it done or do you intend to do more?

CattleProdigy commented:
> @CattleProdigy: I'm a little confused of the state right now because of the rebase and the fixup commit messages. Is it done or do you intend to do more?

Fixups are intended to show the reviewer precisely the changes they asked for. Before merging, the fixups are squashed into whichever commits they refer to. The alternative is to either make additional regular commits or to amend-commit the changes to the original commit and force push the branch. Making additional commits muddies the history and force pushes requires make it harder to see what I change since the last review. 

Shortly, I'll squash these commits, force push the branch and remove the 'WIP' from the title.

Andlon commented:
@CattleProdigy: thanks for finalizing this! I'm not sure if your suggested PR workflow is ideal from the perspective of a reviewer... It seems preferable to me to retain the history, since right now I cannot see the changes you made *after* my review (I think?). So it wouldn't be clear to me if something else also changed in the process. I think multiple smaller commits with clear commit messages are preferable to me, at least.

@sebcrozet: can you approve the CI run, please?

CattleProdigy commented:
> @CattleProdigy: thanks for finalizing this! I'm not sure if your suggested PR workflow is ideal from the perspective of a reviewer... It seems preferable to me to retain the history, since right now I cannot see the changes you made _after_ my review (I think?). So it wouldn't be clear to me if something else also changed in the process. I think multiple smaller commits with clear commit messages are preferable to me, at least.
> 
> @sebcrozet: can you approve the CI run, please?

@Andlon this is something teams debate about and choose for themselves. There isn't a clear right answer. Some teams like to have the merged history reflect the entire decision-making process and some want the history to reflect the _effective_ evolution of the code, omitting any dead-ends. I generally prefer the latter, because when look at the history of `dev` branch e.g., I don't really care that at one point we considered naming the function `push_mat` or that bugs were fixed before merging. Ultimately, this is not my project, so I'll adhere to whatever process you all use. 

Anyways, you can still see the changes by expanding any of those threads and then clicking the fixup SHA1s I posted. You'll have to trust that I did, in fact, apply those fixups.

CattleProdigy commented:
> > @CattleProdigy: thanks for finalizing this! I'm not sure if your suggested PR workflow is ideal from the perspective of a reviewer... It seems preferable to me to retain the history, since right now I cannot see the changes you made _after_ my review (I think?). So it wouldn't be clear to me if something else also changed in the process. I think multiple smaller commits with clear commit messages are preferable to me, at least.
> > @sebcrozet: can you approve the CI run, please?
> 
> @Andlon this is something teams debate about and choose for themselves. There isn't a clear right answer. Some teams like to have the merged history reflect the entire decision-making process and some want the history to reflect the _effective_ evolution of the code, omitting any dead-ends. I generally prefer the latter, because when look at the history of `dev` branch e.g., I don't really care that at one point we considered naming the function `push_mat` or that bugs were fixed before merging. Ultimately, this is not my project, so I'll adhere to whatever process you all use.
> 
> Anyways, you can still see the changes by expanding any of those threads and then clicking the fixup SHA1s I posted. You'll have to trust that I did, in fact, apply those fixups.

Here's a short post about the style of workflow I used here if you're curious: https://dev.to/koffeinfrei/the-git-fixup-workflow-386d

Andlon commented:
@CattleProdigy: thanks for the explanation and the article! I can see how this workflow can work well in an integrated team that is working closely together, but it assumes that everyone (including the reviewer) is on board with the workflow. My impression here is that this doesn't translate well to an open source project, since

1. it assumes the reviewer is familiar with the workflow
2. since the history effectively disappears and all the tools GitHub provides for change tracking are rendered useless, it makes it necessary for the reviewer to put a *lot* of trust into the fact that the PR writer actually did *exactly* what was discussed. Otherwise the reviewer would need to completely review the PR again. In fact, as a reviewer it's completely unclear what was *actually* done in the end (besides looking up the individual fixup commits, but this is very time consuming, and still assumes that the PR writer posted (all) the correct fixups)

It is not quite clear to me what is gained by having a "clean history" when the drawbacks are so severe. The main problem with an "unclean history" is that individual commits likely don't all pass CI, so that `git bisect` might fail later on. However, for this a "squash-and-merge" works much better, since it doesn't break the main review tools at our disposal (GitHub). It also gives you a "clean history" in the sense that the main commit message is "clean", but still includes the individual fixup commit messages in the long-form text. Note that `nalgebra` doesn't currently use "squash-and-merge", however.

Andlon commented:
@sebcrozet: this should be ready for merging now! Changes are local to `nalgebra-sparse` and include no breaking changes (just added API).

sebcrozet commented:
Thank you @CattleProdigy for this PR, and @Andlon for the review!
